### PR TITLE
Upgrade Lepton library, handle its C++11 requirement

### PR DESCRIPTION
**Pull request by giacomofiorin**
**Summary**

The latest version of Lepton from [the OpenMM repository](https://github.com/openmm/openmm) uses C++11 features.  This PR upgrades the copy of Lepton in LAMMPS and handles the new requirement by a CMake recipe or updated conventional make flags for `USER-COLVARS`.

**Author(s)**

@giacomofiorin 
The Lepton library is written by @peastman

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

For CMake-based builds, Lepton is turned on automatically if the C++ level is high enough, and disables it otherwise. For conventional make, some changes to the Colvars `Makefile.common` allow to turn off Lepton by setting the environment variable `ENABLE_LEPTON=no`.

**Implementation Notes**

`USER-COLVARS` now uses a separate CMake recipe file included similarly to other packages. Inside it, there is a conditional:
```
if(DEFINED CMAKE_CXX_STANDARD)
    # The condition below should match most modern C++ standards
    if((${CMAKE_CXX_STANDARD} GREATER 0) AND (${CMAKE_CXX_STANDARD} LESS 70))
...
```
Per the CMake doc, the acceptable values are 98, 11, 14, 17, etc. and the test above should work for some time without edits, regardless of the C++ level set globally for LAMMPS.  Suggestions to improve it are nonetheless welcome.

The list of `include` for user packages in CMakeLists.txt is now sorted alphabetically.

**Post Submission Checklist**

- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system

### CODE REVIEW

**Comment by giacomofiorin** [comment]
Note: I have not added yet a CMake flag to manually disable Lepton even when C++11 requirements are fulfilled. The current flag `ENABLE_LEPTON` is set to ON regardless of user input.

I'm hesitant to create a package for Lepton just to have a `-DPKG_...` flag.  Alternatively, `-DWITH_LEPTON` may work, but the corresponding `WITH_*` flags are used for general-purpose system libraries, not for a scientific library.

**Comment by akohlmey** [comment]
> Note: I have not added yet a CMake flag to manually disable Lepton even when C++11 requirements are fulfilled. The current flag `ENABLE_LEPTON` is set to ON regardless of user input.
> 
> I'm hesitant to create a package for Lepton just to have a `-DPKG_...` flag. Alternatively, `-DWITH_LEPTON` may work, but the corresponding `WITH_*` flags are used for general-purpose system libraries, not for a scientific library.

consistent with other features and packages in LAMMPS, you could define an `option(COLVARS_LEPTON)` with default set to on. Then users could turn it on or off with `-DCOLVARS_LEPTON=<boolean>`. By using both COLVARS and LEPTON in the option flag, it is evident, that this is a suboption to the USER-COLVARS package. Making LEPTON as separate package would not make much sense indeed.

File: cmake/Modules/Packages/USER-COLVARS.cmake [hunk 1]
```diff
@@ -4,22 +4,36 @@ if(PKG_USER-COLVARS)
...
-    if((${CMAKE_CXX_STANDARD} GREATER 0) AND (${CMAKE_CXX_STANDARD} LESS 70))
-      set(ENABLE_LEPTON ON)
-      set(LEPTON_DIR ${LAMMPS_LIB_SOURCE_DIR}/colvars/lepton)
-      file(GLOB LEPTON_SOURCES ${LEPTON_DIR}/src/[^.]*.cpp)
-      add_library(lepton STATIC ${LEPTON_SOURCES})
-      target_include_directories(lepton PRIVATE ${LEPTON_DIR}/include)
+  # Build Lepton by default
+  set(COLVARS_LEPTON_DEFAULT ON)
+  # but not if C++11 is disabled per user request
+  if(DEFINED DISABLE_CXX11_REQUIREMENT)
```
Comment (junghans) [inline cmake/Modules/Packages/USER-COLVARS.cmake]: `DEFINED` is not good, because it could be defined as `FALSE`.

Comment (junghans) [reply cmake/Modules/Packages/USER-COLVARS.cmake]: No wait, that is ok.

**Comment by giacomofiorin** [comment]
Just added the flag in 2c9ae7a. The default value is taken based on the value of `DISABLE_CXX11_REQUIREMENT`, if defined, instead of `CMAKE_CXX_STANDARD`.  I presume the former flag will be used portably..

**Comment by giacomofiorin** [comment]
@junghans You deleted your comment, so I'll just reply here because I think it's informative.  I check for `if(DEFINED DISABLE_CXX11_REQUIREMENT)` as I expect that this option may be removed at some point in the future, at which point the entire LAMMPS can be assumed to be C++11 compliant.

**Comment by junghans** [comment]
> @junghans You deleted your comment, so I'll just reply here because I think it's informative. I check for `if(DEFINED DISABLE_CXX11_REQUIREMENT)` as I expect that this option may be removed at some point in the future, at which point the entire LAMMPS can be assumed to be C++11 compliant.

Yeah, I saw you check for both `DEFINED` and the boolean value, so that should be good.

**Pull request merged**